### PR TITLE
DAOS-10942 utils: chmod should ignore unsupported bits

### PR DESCRIPTION
**Pull request by chowes**
The setuid, setgid, and sticky bit can cause fatal errors when the datamover tool sets file permissions after copying a file, since these are not supported by DFS. We can just ignore this bit when calling dfs_chmod.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/cmd/daos/filesystem.go
Comment (mjmac) [inline src/control/cmd/daos/filesystem.go:50]: minor: With the assumption that there may be more areas of POSIX noncompliance found, it might make sense to generalize this a little to `IgnoreUnsupported`, with the flag being `--ignore-unsupported`, and the help text being "ignore unsupported filesystem features when copying to DFS". From a maintainability standpoint, I could see this being a little bit of a moving target otherwise. From a UX standpoint, we probably wouldn't want to ask the user to have to figure out which combinations of unsupported things they want to ignore, if we need to add more things that are known not to work but shouldn't be fatal errors.

File: src/include/daos_fs_sys.h
Comment (mjmac) [inline src/include/daos_fs_sys.h:188]: Likewise here... We're making a public API change, so it would probably be best to generalize, as I suspect that callers won't really care about having fine-grained control over which unsupported features should be ignored.

File: src/utils/daos_hdlr.c
Comment (mjmac) [inline src/utils/daos_hdlr.c:531]: As this is an internal function, I think it's OK to keep the specific parameters especially in light of the inout param used for reporting.

File: src/control/cmd/daos/filesystem.go
Comment (chowes) [reply src/control/cmd/daos/filesystem.go:50]: Seems reasonable, I feel like that argument was pretty clunky anyway. This should be more usable.

File: src/include/daos_fs_sys.h
Comment (chowes) [reply src/include/daos_fs_sys.h:188]: Done.

**Review by ashleypittman — DISMISSED** [review]

**Comment by mjmac** [comment]
Latest revision LGTM, but I've added @mchaarawi for another client reviewer.

File: src/include/daos_fs_sys.h
Comment (mchaarawi) [inline src/include/daos_fs_sys.h:193]: this is a major version API change (it breaks the existing API); so not acceptable as is.

**Review by mchaarawi — CHANGES_REQUESTED** [review]
This can't land as is because of the API change.
@daltonbohning mpifileutils uses this API, no?
We would need to update mfu to use this new API but also add proper version checks (which this PR does not bump) to work with the existing API (2.4) and the new API (2.5).

alternatively, it's probably easier to just add a new API instead of breaking the API

File: src/include/daos_fs_sys.h
Comment (daltonbohning) [reply src/include/daos_fs_sys.h:193]: Right. For example, third-party mpifileutils uses this function. Which would require a recompile. And, unlikely but possible, some unknown user application.

**Comment by mjmac** [comment]
> This can't land as is because of the API change.
> 
> alternatively, it's probably easier to just add a new API instead of breaking the API
>
Another option, perhaps, would be to return to the original implementation that didn't change the API and just emitted a warning for unsupported bits. I don't feel strongly either way, but I gather that the API change was driven by some feedback that the behavior should be configurable.

File: src/utils/daos_hdlr.c
Comment (daltonbohning) [inline src/utils/daos_hdlr.c:550]: We could avoid an API change and the need for a new function by handling the bits here instead of within `dfs_sys_chmod`. The benefit to handling here is that we could unset the bits _**before**_ calling `dfs_sys_chmod`, avoiding the extra call.
The downside is that if other applications use `dfs_sys_chmod`, they potentially need similar logic.

@mchaarawi @ashleypittman What do you think about handling here vs in the `dfs_sys` API?

**Review by daltonbohning — CHANGES_REQUESTED** [review]

**Review by daltonbohning — COMMENTED** [review]
I don't think you have visibility into the Jenkins CI, but will you please include this line in future commit messages so extra tests are ran?
```
Features: daos_fs_copy
```

**Comment by mchaarawi** [comment]
> > This can't land as is because of the API change.
> > alternatively, it's probably easier to just add a new API instead of breaking the API
> 
> Another option, perhaps, would be to return to the original implementation that didn't change the API and just emitted a warning for unsupported bits. I don't feel strongly either way, but I gather that the API change was driven by some feedback that the behavior should be configurable.

right, so just emitting a warning has its downside where it's not always visible to the caller that what they did is not supported. But this does not mean that we can just break the API without any regards to proper versioning and handling for API & ABI compatibility!
This PR requires several things to be acceptable:
1) bump the API major version to X+1
2) patch mpifileutils to have compile time check for the API version (if <= X use old API; if > X use new API). see an example here: https://github.com/hpc/ior/blob/main/src/aiori-DFS.c#L519
3) update RPM packaging for our CI for mpifileutils after 2 lands to MFU.

and yes it's not an easy change to break the API and it shouldn't be..
which is why i suggested the alternative to add a new API function instead of breaking the API:
dfs_sys_chmod_wflags(..., uin32_t flags); where one of the flags can be allow_unsupported bit. and use that function where this functionality is desired.

the other suggestion as Dalton was saying in his comment is to handle the change in where this is called without adding the new API (or breaking the old one).
I do not have a strong opinion on either as long as it's done properly.

**Comment by chowes** [comment]
I agree with Dalton's suggestion, let's just move the error handling outside of dfs_sys_chmod. Other applications would need to update anyway to support the parameter, so I don't think the extra burden of adding the error handling is worth the burden of workout around a breaking API change.

File: src/include/daos_fs_sys.h
Comment (daltonbohning) [inline src/include/daos_fs_sys.h:189]: To undo these unnecessary format changes, probably can just do
```
git checkout origin/master -- src/include/daos_fs_sys.h
```

File: src/client/dfs/dfs_sys.c
Comment (daltonbohning) [inline src/client/dfs/dfs_sys.c:2]: Minor, but tricky to backout the githook copyright change. I think you have to
1. Manually change back to 2022
2. `git commit --no-verify` to skip the githooks
3. Continue with other changes
4. Make sure your last commit was NOT ran with `--no-verify` so the githooks take affect for subsequent changes

File: src/tests/suite/dfs_sys_unit_test.c
Comment (daltonbohning) [inline src/tests/suite/dfs_sys_unit_test.c:2]: And here...

**Review by daltonbohning — COMMENTED** [review]
Content LGTM - thanks! Minor comments about backing out the no-longer-needed stuff

**Review by daltonbohning — APPROVED** [review]

File: src/utils/daos_hdlr.c
Comment (mchaarawi) [inline src/utils/daos_hdlr.c:546]: Im nor sure i quite understand why a retry is needed here.
so the user has not requested ignore_unsup, but you still want to ignore_unsup?

Comment (chowes) [reply src/utils/daos_hdlr.c:546]: Yeah, the idea is...

with ignore_unsup:
- unset the bits before the call to chmod
- there will be no error, we won't log a warning, and we exit 0

without ignore_unsup:
- the call to chmod fails
- unset the bits and try again
- print a warning and exit non-zero

Some history of the discussion in the bug: https://daosio.atlassian.net/browse/DAOS-10942

Comment (daltonbohning) [reply src/utils/daos_hdlr.c:546]: Looks like with and without `--ignore-unsupported`, the unsupported mode bits are removed. By default, `num_chmod_enotsup` is reported at the end. But with `--ignore-unsupported` it's silent.

Maybe to avoid a retry we could do something like this upfront?
```
if (!ignore_unsup && mode & (S_ISVTX | S_ISGID | S_ISUID))
    (*num_chmod_enotsup)++;
mode &= ~(S_ISVTX | S_ISGID | S_ISUID);
```

Comment (chowes) [reply src/utils/daos_hdlr.c:546]: That seems reasonable. The only downside I can see is that maybe we eventually support these bits and it's not an error anymore. In that case we'd probably want to update this code anyway though.

Comment (daltonbohning) [reply src/utils/daos_hdlr.c:546]: Yeah, I think we'd want to update this anyway if we support in the future, and if at all I expect it would be far off

Comment (mchaarawi) [reply src/utils/daos_hdlr.c:546]: if you are going to update this please also merge latest master to your branch as it is lagging behind a bit now.

Comment (chowes) [reply src/utils/daos_hdlr.c]: Shoot, my attempt to rebase brought in a bunch of random commits. Sorry, I'm not used to github yet.

Comment (daltonbohning) [reply src/utils/daos_hdlr.c]: The flow is usually
```
git checkout my_branch
git fetch
git merge origin/master
```

Comment (chowes) [reply src/utils/daos_hdlr.c]: Yeah, I'm talking to Mike and Jeff and they are saying that github doesn't handle rebases well. Okay let me see if I can fix it. If not, I'll create a new PR.

Comment (daltonbohning) [reply src/utils/daos_hdlr.c]: Did you use `rebase` or `merge`? `rebase` is trickier because it re-writes history (i.e. commits end up with different hashes). It's better to use `merge` after people have started reviewing so the history of commits is preserved.
Another common thing is if you do `git merge master`, your local `master` might not be up-to-date with the remote `master`. So you either have to do this
```
git checkout master
git fetch
git pull
git checkout my_branch
git merge master
```
or (simpler I think)
```
git checkout my_branch
git fetch
git merge origin/master
```

Comment (chowes) [reply src/utils/daos_hdlr.c]: I did a rebase the first time. I think it should be okay now.

**Review by daltonbohning — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Comment by daltonbohning** [comment]
Single test failure in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-12949/20/testReport/
is existing and unrelated to these changes:
https://daosio.atlassian.net/browse/DAOS-14450

**Review by mchaarawi — APPROVED** [review]

File: src/control/cmd/daos/filesystem.go
Comment (tanabarr) [inline src/control/cmd/daos/filesystem.go:113]: errors.Errorf("...", fargs) should be used to create error from format string

**Review by tanabarr — COMMENTED** [review]
PRs touching go code should probably be reviewed by a go developer before landing

**Comment by daltonbohning** [comment]
> PRs touching go code should probably be reviewed by a go developer before landing

It was reviewed by a Go developer, but just not approved by one.

**Pull request merged**
